### DISCUSSION TITLE
i2c.writeto() requires buffer protocol

### DISCUSSION THREAD

**Discussion by nick6502** [micropython/micropython#11347] [RP2040 / Pico] [answered]
I'm new to microPython and the Pi Pico board, attempting to learn some python by programming a synthesizer IC. 
In trying to send data via I2C, I get an error saying that the line with the i2C.writeto() call requires buffer protocol. But I see examples elsewhere in which the code just has the call, address, and data which is what I'm doing. I started with a more complex programming with bit manipulation in the arguments, function calls and so on but I kept reducing complexity until I got to this minimal code and still get the same error.

I'm using Thonny ...
Program:
import machine
import utime
sda=machine.Pin(8)
scl=machine.Pin(9)
i2c=machine.I2C(0,sda=sda, scl=scl, freq=400000)

adr = 26
dat = 255

print(type(adr))
print(type(dat))

print("Hello")
i2c.writeto(adr, dat)

Output:

>>> %Run -c $EDITOR_CONTENT
<class 'int'>
<class 'int'>
Hello
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
TypeError: object with buffer protocol required
>>> 

Any ideas?

Thanks,

Nick

**Comment by davefes** [comment]
Try a power on reset and see if it happens on the first call.

**Comment by bixb922** [comment]
dat could for example be a bytearray but not an integer

**Comment by mcauser** [comment, marked as answer]
Simply update this:
```python
i2c.writeto(adr, dat)
```

To this:
```python
i2c.writeto(adr, bytearray([dat]))
```

Or alternatively, keep a bytearray variable for the data, then update it's first byte when it changes.
```python
buf = bytearray(1)
buf[0] = 255
i2c.writeto(adr, buf)
...
buf[0] = 223
i2c.writeto(adr, buf)
```

**Reply by nick6502** [reply]
Thanks for the responses.  I eventually found some online code that I copied and the error went away. I had to put my i2c.write() call inside another function which provided the necessary bytearray buffer. Not sure why the address argument didn't have to be buffered, but OK ...

def i2c_write_byte(addr, reg, v):
    buf = bytearray(2)
    buf[0] = reg & 0xFF
    buf[1] = v & 0xFF
    i2c.writeto(addr, buf)

Eventually I'll figure out how to paste in a block of code with proper formatting.

Nick

**Reply by mcauser** [reply]
To output formatted code, wrap the code in 3x backticks ` at the top and bottom. Optionally with a language at the top.
Like this, but backticks instead of single quotes.

'''python
print('hello')
'''

**Reply by nick6502** [reply]
OK, I'm going to give it a try. First I had to look up what a backtick is.  (Below the tilde)

```python
from machine import Pin
from utime import sleep

pin = Pin(25, Pin.OUT)

print("LED starts flashing...")
while True:
    pin.toggle()
    sleep(1) # sleep 1sec ```

Ah, I can already see from the Preview that it works.  Now I just have to remember ...

Nick

**Comment by gregwa1953** [comment]
I'm having a similar problem however, the suggested fix doesn't work.

Micropython version: MicroPython v1.20.0-50-g786013d46 on 2023-05-04; Raspberry Pi Pico W with RP2040

When I tried replacing the i2c.writeto(adr,dat) with i2c.writeto(adr,bytearray([dat]))
I got back an error " TypeError: can't convert bytearray to int"

Module ssd1306 line 116 - self.i2c.writeto(self.addr, self.temp) # self.temp=bytearray(b'\x80\xae')
Module bmp180 line 48 - self.chip_id = self._bmp_i2c.readfrom_mem(_bmp_addr, 0xD0, 2)

These errors did not occur under 1.19.1-192 or any previous versions of MicroPython on the Pico/Pico-w

**Comment by Prof-C4ar7es** [comment]
I noticed, at least for the micropython (Raspberry pico) version that there is a command i2c.writevto(addr, buffer) command - that seems to work for me.

Not sure if this helps 

 def write_data(self, buf):
        self.write_list[1] = buf
        print(buf)
        self.i2c.writevto(self.addr, self.write_list)
